oauth2_http/javatests/com/google/auth/oauth2/StsRequestHandlerTest.java

```diff
@@ -0,0 +1,311 @@
+                requestHandler.exchangeToken();
+              }
+            });
+
+    assertThat(e.getErrorCode()).isEqualTo(INVALID_REQUEST);
+    assertThat(e.getErrorDescription()).isNull();
+    assertThat(e.getErrorUri()).isNull();
+  }
+
+  @Test
+  public void exchangeToken_withOptionalParams_throwsException() throws IOException {
+    StsTokenExchangeRequest stsTokenExchangeRequest =
+        StsTokenExchangeRequest.newBuilder(CREDENTIAL, SUBJECT_TOKEN_TYPE).build();
+
+    final StsRequestHandler requestHandler =
+        StsRequestHandler.newBuilder(
+                TOKEN_URL, stsTokenExchangeRequest, transport.createRequestFactory())
+            .build();
+
+    transport.addResponseErrorSequence(
+        buildHttpResponseException(INVALID_REQUEST, ERROR_DESCRIPTION, ERROR_URI));
+
+    OAuthException e =
+        assertThrows(
+            OAuthException.class,
+            new ThrowingRunnable() {
+              @Override
+              public void run() throws Throwable {
+                requestHandler.exchangeToken();
+              }
+            });
+
+    assertThat(e.getErrorCode()).isEqualTo(INVALID_REQUEST);
+    assertThat(e.getErrorDescription()).isEqualTo(ERROR_DESCRIPTION);
+    assertThat(e.getErrorUri()).isEqualTo(ERROR_URI);
+  }
+
+  @Test
+  public void exchangeToken_ioException() {
+    StsTokenExchangeRequest stsTokenExchangeRequest =
+        StsTokenExchangeRequest.newBuilder(CREDENTIAL, SUBJECT_TOKEN_TYPE).build();
+
+    final StsRequestHandler requestHandler =
+        StsRequestHandler.newBuilder(
+                TOKEN_URL, stsTokenExchangeRequest, transport.createRequestFactory())
+            .build();
+
+    IOException e = new IOException();
+    transport.addResponseErrorSequence(e);
+
+    IOException thrownException =
+        assertThrows(
+            IOException.class,
+            new ThrowingRunnable() {
+              @Override
+              public void run() throws Throwable {
+                requestHandler.exchangeToken();
+              }
+            });
+    assertThat(thrownException).isEqualTo(e);
+  }
+
+  public HttpResponseException buildHttpResponseException(
+      String error, @Nullable String errorDescription, @Nullable String errorUri)
+      throws IOException {
+    ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+
+    JsonGenerator jsonGenerator =
+        OAuth2Utils.JSON_FACTORY.createJsonGenerator(
+            byteArrayOutputStream, Charset.defaultCharset());
+
+    jsonGenerator.writeStartObject();
+    jsonGenerator.writeFieldName("error");
+    jsonGenerator.writeString(error);
+
+    if (errorDescription != null) {
+      jsonGenerator.writeFieldName("error_description");
+      jsonGenerator.writeString(errorDescription);
+    }
+    if (errorUri != null) {
+      jsonGenerator.writeFieldName("error_uri");
+      jsonGenerator.writeString(errorUri);
+    }
+    jsonGenerator.writeEndObject();
+    jsonGenerator.close();
```

Comment:
It may be easier to use [GenericJson](https://googleapis.dev/java/google-http-client/latest/index.html?com/google/api/client/json/GenericJson.html).

```java
GenericJson genericJson = new GenericJson();
genericJson.setFactory(OAuth2Utils.JSON_FACTORY);
String content = genericJson
    .set("foo", "bar")
    .set("asdf", "qwer")
    .toString();
```

Reply:
Done.